// components/Buttons/TaskControlButton.jsx
import React from 'react';
import AddButton from "components/Buttons/AddButton";
import EditIcon from '@material-ui/icons/Edit';
import DeleteIcon from '@material-ui/icons/Delete';
import RedoIcon from '@material-ui/icons/Redo';
import AccessTimeIcon from '@material-ui/icons/AccessTime';
import AddIcon from '@material-ui/icons/Add';

export const EditButton = ({onClick, label, children}) => {
    return (
        <AddButton title={label} aria-label={label} onClick={onClick}>
            <EditIcon/>{children}
        </AddButton>
    );
};

export const DeleteButton = ({onClick, label, children}) => {
    return (
        <AddButton title={label} aria-label={label} onClick={onClick}>
            <DeleteIcon/> {children}
        </AddButton>
    );
};

export const CheckTimeButton = ({onClick, label, children}) => {
    return (
        <AddButton title={label} aria-label={label} onClick={onClick}>
            <AccessTimeIcon/> {children}
        </AddButton>
    );
};

export const ToTaskButton = ({taskId, label, children}) => {
    return (
        <a rel="noopener noreferrer" target="_blank" href={`https://cp.webpractik.ru/workgroups/group/413/tasks/task/view/${taskId}/`}>
            <AddButton title={label} aria-label={label}>
                <RedoIcon/> {children}
            </AddButton>
        </a>
    );
};


export const NewTask = ({onClick, label, children}) => {
    return (
        <AddButton title={label} aria-label={label} onClick={onClick}>
            <AddIcon/> {children}
        </AddButton>
    );
};

export const NewCheck = ({onClick, label, children}) => {
    return (
        <AddButton title={label} aria-label={label} onClick={onClick}>
            <AccessTimeIcon/> {children}
        </AddButton>
    );
};
