// components/Forms/CreatePlanItem/CreatePlanItem.js
import React, {Component} from 'react';
import Button from '@material-ui/core/Button';
import CreateFormSelect from 'components/Forms/CreatePlanItem/Selects/CreateFormSelect';
import CreateFormTextField from 'components/Forms/CreatePlanItem/TextFields/CreateFormTextField';
import ProjectSelect from 'components/Forms/CreatePlanItem/ProjectSelect/ProjectSelect';
import TaskSelect from 'components/Forms/CreatePlanItem/TaskSelect/TaskSelect';
import H2 from 'components/ui/H2/H2';
import ZONE_TYPE from 'config/zoneTypes.config';
import { Formik } from 'formik';
import projectList from 'startData/projects';
import './CreatePlanItem.css';
import PeopleInput from "../../Fields/AutoCompleteInput/Peoples";


class CreatePlanItem extends Component {

    static defaultProps = {
        params: {},
        planItemsCollection: {}
    };

    submit = (data) => {
        let params = {
            projectName: data.project || 'Проект не указан',
            title: data.task || '',
            date: this.props.params.date,
            userId: this.props.params.userId || '',
            zone: data.zone || ZONE_TYPE.PRODUCTIVE,
            time: data.time || '15',
            additionDesc: data.desc || '',
            timeChecked: 0,
            type: this.props.params.type,
            end: this.props.params.end,
            start: this.props.params.start
        };

        this.props.successCallback(params)
            .then( result => {
                return new Promise((resolve, reject) => {
                    if (result.status) {
                        resolve(result);
                        this.props.cancel();
                    } else {
                        // TODO: Вывод ошибок в форме
                    }
                });
            },
                err => {
                    // TODO: Вывод ошибок в форме
                }
            );

    };

    cancel = () => {
        this.props.cancelCallback()
            .then( result => {
                    return new Promise((resolve, reject) => {
                        if (result.status) {
                            resolve(result);
                            this.props.cancel();
                        } else {
                            // TODO: Вывод ошибок в форме
                        }
                    });
                },
                err => {
                    // TODO: Вывод ошибок в форме
                }
            );
    };

    render() {
        return (
            <div className={'create-plan-item'}>
               <Formik
                   initialValues={{
                       project: this.props.params.projectName || '',
                       task: this.props.params.title || '',
                       desc: this.props.params.additionalDesc || '',
                       zone: this.props.params.zone || '',
                       time: this.props.params.time || 15
                   }}

                   onSubmit={(values, { setSubmitting }) => {
                       this.submit(values, setSubmitting);
                   }}

                   render={(props) => (
                       <form onSubmit={props.handleSubmit}>
                           <H2>Создать бронь</H2>
                           <ProjectSelect
                               id={'project'}
                               name={'project'}
                               label={'Проект'}
                               items={projectList.projects}
                               onChange={props.handleChange}
                               onBlur={props.handleBlur}
                               value={props.values.project}
                           />
                           <TaskSelect
                               id={'task'}
                               name={'task'}
                               label={'Задача'}
                               items={projectList.tasks}
                               onChange={props.handleChange}
                               onBlur={props.handleBlur}
                               value={props.values.task}
                           />
                            <CreateFormTextField
                                id={'desc'}
                                name={'desc'}
                                label={'Дополнительное описание'}
                                value={props.values.desc}
                                onChange={props.handleChange}
                                onBlur={props.handleBlur}
                            />
                           <CreateFormTextField
                               id={'time'}
                               name={'time'}
                               label={'Время в минутах'}
                               value={props.values.time}
                               onChange={props.handleChange}
                               onBlur={props.handleBlur}
                           />
                           <Button type={'submit'} color="primary" variant="contained">Сохранить</Button>
                           <span>&nbsp;&nbsp;&nbsp;&nbsp;</span>
                           <Button color="secondary" onClick={this.cancel} variant="contained">Отмена</Button>
                       </form>
                   )}
               >
               </Formik>
            </div>
        );
    }
}

export default CreatePlanItem;
