// components/Forms/CreatePlanItem/CreatePlanItem.js
import React, {Component} from 'react';
import MenuItem from '@material-ui/core/MenuItem';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import CreateFormSelect from 'components/Forms/Selects/CreateFormSelect';
import { Formik } from 'formik';
import projectList from 'startData/projects';
import './CreatePlanItem.css';

class CreatePlanItem extends Component {
    render() {
        return (
            <div className={'create-plan-item'}>
               <Formik
                   initialValues={{
                       project: null,
                       task: '',
                       desc: '',
                       employer: '',
                       zone: '',
                       time: ''
                   }}

                   render={(props) => (
                       <form onSubmit={props.handleSubmit}>
                           <CreateFormSelect
                               id={'project'}
                               name={'project'}
                               items={projectList.projects}
                               onChange={props.handleChange}
                               onBlur={props.handleBlur}
                               value={props.values.project}
                           />
                           <br/>
                            <TextField
                                id={'desc'}
                                name={'desc'}
                                label={'Описание задачи'}
                                value={props.values.desc}
                                onChange={props.handleChange}
                                onBlur={props.handleBlur}
                                margin="normal"
                                variant="outlined"
                            />
                           <Button variant="contained">Создать</Button>
                       </form>
                   )}
               >
               </Formik>
            </div>
        );
    }
}

export default CreatePlanItem;
