// components/DaysList/DaysList.js
import React, {Component} from 'react';
import AppDate from 'entities/date';
import {observer} from "mobx-react";
import FullCalendar from '@fullcalendar/react'
import dayGridPlugin from '@fullcalendar/daygrid';
import bootstrapPlugin from '@fullcalendar/bootstrap';
import interactionPlugin, { Draggable } from '@fullcalendar/interaction'
import resourceTimeGridPlugin from '@fullcalendar/resource-timegrid';
import '@fullcalendar/core/main.css';
import '@fullcalendar/daygrid/main.css';
import '@fullcalendar/timegrid/main.css';

const events = [
  {title: 'Трудиться как раб на галере', start: '2019-09-19 10:00:00', end: '2019-09-19 12:00:00'},
  {title: 'Попить чаю', start: '2019-09-19 13:00:00', end: '2019-09-19 13:15:00'},
];

const DaysList = observer(
    class DaysListClass extends Component {
        constructor(props) {
            super(props);
            this.appDateInstance = new AppDate();
        }

        render() {
            return (
              <>
              <FullCalendar
                defaultView='timeGridWeek'
                editable
                minTime={"10:00:00"}
                maxTime={"15:00:00"}
                slotDuration={"0:15"}
                events={events}
                locale={'ru'}
                weekends={this.props.daysCount !== "7" ? false : true}
                firstDay={1}
                slotLabelFormat={{
                  hour: '2-digit',
                  minute: '2-digit',
                  hour12: false,
                  omitZeroMinute: false,
                  meridiem: false
                }}
                plugins={[ dayGridPlugin, interactionPlugin, resourceTimeGridPlugin  ]}
              >
                <Draggable/>
              </FullCalendar>
              </>
            );
        }
    }
);

export default DaysList;
