// components/HOC/withThemeParams.jsx
import React, {Component} from 'react';
import ThemeModel from 'RestModels/ThemeModel';
import {EmptyFunc} from 'services/helper';


const withThemeParams = function (WrappedComponent, theme) {
    return class FilterBy extends Component {

        static defaultProps = {
            onError: EmptyFunc,
            onSuccess: EmptyFunc
        };

        state = {
            themeParams: null
        };

        getParams = async () => {
            const model = new ThemeModel();

            try {
                const result = await model.getAll(theme);
                this.setState({themeParams: result}, () => {
                    this.props.onSuccess()
                });
            } catch (e) {
                this.props.onError(e.message);
            }
        };

        componentDidMount() {
            this.getParams().then(() => {});
        }

        render() {
            if (!this.state.themeParams) return false;

            return (
                <WrappedComponent {...this.props} {...this.state.themeParams}/>
            );
        }
    }
};


export default withThemeParams;
