// RestModels/UsersModel.js
import $ from 'jquery';
import {siteOptions} from "config/app.config";

export default class EventModel {
    static defaultOptions = {
        ...siteOptions,
        SHOW_INACTIVE_USERS: 'N',
        SHOW_EXTRANET_USERS: 'ALL',
        EX_GROUP: '',
        nt: '#NAME# #LAST_NAME#',
        sl: 'N',
        SHOW_USERS: 'Y'
    };

    constructor(options) {
        this.options = {...EventModel.defaultOptions, ...options}
    }

    async getAll(themeName) {
        return new Promise( (resolve, reject) => {
            $.get('/bitrix/components/bitrix/intranet.user.selector.new/ajax.php').done((result) => {
                resolve(result);
            })
                .fail((err) => {
                    reject(err);
                })
        })
    }

    async findByName(string) {
        const optionsToSend = this.options;

        optionsToSend.MODE = 'SEARCH';
        optionsToSend.SEARCH_STRING = string;

        $.get('/bitrix/components/bitrix/intranet.user.selector.new/ajax.php', optionsToSend).done((result) => {
            resolve(result);
        })
            .fail((err) => {
                reject(err);
            })
    }
}
