// components/Notifiers/SnackNotifier/index.jsx
import React, {Component} from 'react';
import Slide from '@material-ui/core/Slide';
import Snackbar from '@material-ui/core/Snackbar';
import PropTypes from 'prop-types';

function TransitionUp(props) {
    return <Slide {...props} direction="up"/>;
}

class SnackNotifier extends Component {

    static propTypes = {
        open: PropTypes.bool,
        status: PropTypes.string,
        message: PropTypes.string,
        hide: PropTypes.func
    };

    render() {
        return (
            <Snackbar
                anchorOrigin={{
                    vertical: 'bottom',
                    horizontal: 'center',
                }}
                className={`snack-notify snack-notify--${this.props.status}`}
                open={this.props.open}
                autoHideDuration={5000}
                onClose={this.props.hide}
                TransitionComponent={TransitionUp}
                onClick={this.props.hide}
                ContentProps={{
                    'aria-describedby': 'message-id',
                }}
                message={<span id="message-id">{this.props.message}</span>}
            />

        );
    }
}

export default SnackNotifier;
