// components/CalendarController/CalendarController.js
import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {observer} from "mobx-react";
import FullCalendar from '@fullcalendar/react'
import dayGridPlugin from '@fullcalendar/daygrid';
import interactionPlugin, {Draggable} from '@fullcalendar/interaction'
import resourceTimeGridPlugin from '@fullcalendar/resource-timegrid';
import {CalendarApiItem, CalendarApiDistributor} from 'entities/calendarApiDistributor';
import {EmptyFunc} from 'services/helper';
import '@fullcalendar/core/main.css';
import '@fullcalendar/daygrid/main.css';
import '@fullcalendar/timegrid/main.css';
import './DaysList.css';

const ApiDistributor = CalendarApiDistributor.getInstance();

/**
 * Facade for FullCalendar component. Configuration and callbacks
 * @type {{defaultProps, new(*=): DaysListClass, calendarRef: *, calendarApi: *, new<P, S>(props: Readonly<P>): DaysListClass, new<P, S>(props: P, context?: any): DaysListClass, prototype: DaysListClass}}
 */
const CalendarController = observer(
    class DaysListClass extends Component {
        constructor(props) {
            super(props);
            this.calendarRef = React.createRef()
        }

        static defaultProps = {
            planItems: [],
            calendarName: 'calendar_dev',
            hours: 2,
            showHeader: false,
            eventClickHandler: EmptyFunc,
            eventDropHandler: EmptyFunc,
            eventResizeHandler: EmptyFunc,
            onSelectTime: EmptyFunc,
            onUnselectTime: EmptyFunc,
            weekChanged: EmptyFunc,
        };

        static propTypes = {
            planItems: PropTypes.arrayOf(PropTypes.object),
            calendarName: PropTypes.string,
            hours: PropTypes.number,
            showHeader: PropTypes.bool,
            eventClickHandler: PropTypes.func,
            eventDropHandler: PropTypes.func,
            eventResizeHandler: PropTypes.func,
            onSelectTime: PropTypes.func,
            onUnselectTime: PropTypes.func,
            weekChanged: PropTypes.func,
        };

        state = {
            eventsModels: [],
            planItems: this.props.planItems,
        };

        componentDidMount() {
            /* Creating link for this instance of CalendarApi and adding it to store */
            let calendarApi = this.calendarRef.current.getApi();
            this.calendarApi = new CalendarApiItem(calendarApi);
            ApiDistributor.addApi(this.calendarApi, this.props.calendarName);
        }

        withCalendarApi = (func) => {
            return (event) => {
                event.calendarApi = this.calendarApi;
                event.calendarCode = this.props.calendarCode;
                console.log(event);
                func(event);
            }
        };

        render() {
            const {daysCount, planItems, hours, showHeader} = this.props;
            const showWeekend = (daysCount === "7");

            return (
                <React.Fragment>
                    <FullCalendar
                        ref={this.calendarRef}
                        defaultView='timeGridWeek'
                        editable
                        selectable
                        droppable
                        navLinks
                        minTime={"10:00:00"}
                        maxTime={`${10 + hours}:00:00`}
                        slotDuration={"0:15"}
                        allDayMaintainDuration
                        events={planItems}
                        locale={'ru'}
                        weekends={showWeekend}
                        firstDay={1}
                        snapDuration={"0:15"}
                        allDaySlot={false}
                        columnHeader={showHeader}
                        header={showHeader ? {left: 'title', center: '', right: 'today prev,next'} : false}
                        slotLabelFormat={{
                            hour: '2-digit',
                            minute: '2-digit',
                            hour12: false,
                            omitZeroMinute: false,
                            meridiem: false
                        }}
                        plugins={[dayGridPlugin, interactionPlugin, resourceTimeGridPlugin]}
                        eventResize={this.withCalendarApi(this.props.eventResizeHandler)}
                        select={this.withCalendarApi(this.props.onSelectTime)}
                        unselect={this.withCalendarApi(this.props.onUnselectTime)}
                        eventClick={this.withCalendarApi(this.props.eventClickHandler)}
                        eventDrop={this.withCalendarApi(this.props.eventDropHandler)}
                        drop={this.withCalendarApi(this.props.eventDropHandler)}
                        navLinkWeekClick={this.withCalendarApi(this.props.weekChanged)}
                        navLinkDayClick={this.withCalendarApi(this.props.weekChanged)}
                    >
                        <Draggable minDistance={1}/>
                    </FullCalendar>
                </React.Fragment>
            );
        }
    }
);

export default CalendarController;
