// components/NavBar/NavBar.jsx
import React, {Component} from 'react';
import Button from '@material-ui/core/Button';
import ButtonGroup from '@material-ui/core/ButtonGroup';
import KeyboardArrowLeftIcon from '@material-ui/icons/KeyboardArrowLeft';
import KeyboardArrowRightIcon from '@material-ui/icons/KeyboardArrowRight';
import Grid from '@material-ui/core/Grid';
import {getDaysInMilliseconds, EmptyFunc} from 'services/helper';
import './NavBar.css';

class NavBar extends Component {
    static defaultProps = {
        daysCount: 1,
        onNext: EmptyFunc,
        onPrev: EmptyFunc
    };

    state = {
        currentDate: this.props.currentDate || new Date()
    };

    getChildWithNewProps = () => {
        return React.Children.map(this.props.children, child => {
            //console.log(this.state.currentDate)
            return React.cloneElement(child, {
                selectedDate: this.state.currentDate,
                rand: Math.random()
            })
        });
    };

    next = () => {
        let newCurrentDate = this.state.currentDate;
        newCurrentDate = new Date(this.state.currentDate.getTime()+getDaysInMilliseconds(this.props.daysCount));
        this.setCurrentDate(newCurrentDate);
    };

    prev = () => {
        let newCurrentDate = this.state.currentDate;
        newCurrentDate = new Date(this.state.currentDate.getTime()-getDaysInMilliseconds(this.props.daysCount));
        this.setCurrentDate(newCurrentDate);
    };

    setCurrentDate = (newCurrentDate) => {
        this.setState({currentDate: newCurrentDate}, () => {
            this.props.onPrev(newCurrentDate);
        })
    };

    render() {
        return (
            <div className={'nav-panel'}>
                <Grid container justify="flex-end">
                    <Grid className="nav-panel__wrapper" item>
                    <ButtonGroup variant={"outlined"}>
                        <Button onClick={this.prev}>
                            <KeyboardArrowLeftIcon/>
                        </Button>
                        <Button onClick={this.next}>
                            <KeyboardArrowRightIcon/>
                        </Button>
                    </ButtonGroup>
                    </Grid>
                </Grid>
                {this.props.children(this.state.currentDate)}
            </div>
        );
    }
}

export default NavBar;
