// entities/ui.js
import {decorate, observable, action} from "mobx";
import {EmptyPromise} from 'services/helper';

export class DraggableItemEntity {
    index = null;
    drag = false;
    dragStart(index) {
        this.index = index;
        this.drag = true;
    }

    dragStop() {
        this.drag = false;
        let index = this.index;
        this.index = null;
        return index;
    }
}

decorate(DraggableItemEntity, {
   index: observable,
   drag: observable,
   dragStart: action,
   dragStop: action
});

export class ModalItemEntity {
    visible = false;
    type = null;
    params = {};
    successCallback = EmptyPromise;
    cancelCallback = EmptyPromise;

    setParams(params){
        this.params = Object.assign(this.params, params);
    }

    setType(type) {
        this.type = type;
    }

    clearType() {
        this.type = null;
    }

    makeVisible() {
        this.visible = true;
    }

    makeInvisible() {
        this.visible = false;
    }

    toggleVisible() {
        this.visible = !this.visible;
    }

    setSuccessCallback(func) {
        this.successCallback = func;
    }

    setCancelCallback(func) {
        this.cancelCallback = func;
    }

}

decorate(ModalItemEntity, {
    visible: observable,
    type: observable,
    params: observable,
    successCallback: observable,
    cancelCallback: observable,
    setParams: action,
    setType: action,
    clearType: action,
    makeVisible: action,
    toggleVisible: action,
    setSuccessCallback: action,
    setCancelCallback: action
});
