// entities/ui.js

import {decorate, observable, computed, action} from "mobx";
import {EmptyPromise} from 'services/helper';

class DraggableItem {
    index = null;
    drag = false;
    dragStart(index) {
        this.index = index;
        this.drag = true;
    }

    dragStop() {
        this.drag = false;
        let index = this.index;
        this.index = null;
        return index;
    }
}

export const DraggableItemEntity = decorate(DraggableItem, {
    index: observable,
    drag: observable,
    dragStart: action,
    dragStop: action
});

class ModalItem {
    visible = false;
    type = null;
    params = {};
    successCallback = EmptyPromise;
    cancelCallback = EmptyPromise;

    setParams(params){
        this.params = Object.assign(this.params, params);
    }

    setType(type) {
        this.type = type;
    }

    clearType() {
        this.type = null;
    }

    makeVisible() {
        this.visible = true;
    }

    makeInvisible() {
        this.visible = false;
    }

    toggleVisible() {
        this.visible = !this.visible;
    }

    setSuccessCallback(func) {
        this.successCallback = func;
    }

    setCancelCallback(func) {
        this.cancelCallback = func;
    }

}
export const ModalItemEntity = decorate(ModalItem, {
    visible: observable,
    type: observable,
    params: observable,
    successCallback: observable,
    cancelCallback: observable,
    setType: action,
    setParams: action,
    clearType: action,
    makeVisible: action,
    makeInvisible: action,
    toggleVisible: action,
    setSuccessCallback: action,
    setCancelCallback: action
});
